### ISSUE TITLE
WebGL implementation of calculations

### ISSUE THREAD

**Issue by mdda** [karpathy/convnetjs#13]
(this is a continuation of the discussion started in #11, so it can be 'closed' cleanly)  

> ...  jpcnn, which in turn relies on underscore ...

Hmm - that _is_ a lot of machinery to include 100-200 lines of WebGL.  The least-intrusive method for including the end result (i.e. what the client sees) would be to have a separate `convnet.webgl.min.js` which, if it's there, sets up a webgl flag for the regular `convnet.min.js` to call into - or even overwrite the re-implemented methods themselves.  

On the source side, however, I've got to think there's a more direct way of making the BLAS.js code ready-to-use.  I also think it makes sense to go the BLAS-compatible route, since it's a standard, and one avoids having to continuously re-invent the wheel...  I'll have a poke around for a cleaner set of includes.

All the Best
Martin
:-)

**Comment by bhack** [comment]
@Maratyszcza do you think that it could take a speedup using fusion.js? Do you have done a benchmark against [glmatrix](http://glmatrix.net/)?

**Comment by Maratyszcza** [comment]
@bhack I am not related to [fusion.js](https://code.google.com/p/fusionjs/) project.

If you mean [Furious.js](https://github.com/hpcgarage/furious.js), it does not aim at small matrix manipulations.

**Comment by bhack** [comment]
Yes sorry I meant Furious.js.  So your effort "blis for the web" it was not related to Furious?

**Comment by Maratyszcza** [comment]
@bhack It is intended to be integrated into Furious.js, but it is not there yet. And in any case, it would only help for large matrices/NDArrays

**Comment by ssampang** [comment]
@karpathy Just wondering if there's been any progress on this, as I haven't been able to find much else about it online. 

I'd like to use your deep q learning code to try and train an agent on Go, as most of the necessary additional code is already available online in javascript, but I'm not sure how slow training would be without the speedup of a gpu.

**Comment by waylonflinn** [comment]
I'm thinking about pulling together all the pieces on this topic (GPU acceleration via WebGL) into a standalone library (possibly using the jetpac code as a starting point) for accelerating deep learning and other machine learning applications in the browser.

My hope is to provide a base for libraries like this one (and new entrants) to build on, and (hopefully) get performance gains that approach those found in Caffe. Do people think this could be useful?

**Comment by ssampang** [comment]
@waylonflinn Definitely. GPU acceleration right from the browser could lead to some very cool and useful demo pages that don't need any extra setup to run - not to mention the wide applicability if you didn't tie it to ConvNetJS specifically.

**Comment by waylonflinn** [comment]
@ssampang Glad to hear! I created a repository to track progress. No code there yet, but expect a basic implementation of `gemm` in the next couple of days!

[weblas](https://github.com/waylonflinn/weblas)

I also have an update on the GPU `gemm` code from the Deep Belief SDK. It looks like it doesn't run in Chrome anymore (and possibly has issues in a few other browsers). If anyone is interested in troubleshooting or contributing a fix, you can track that in [issue 57](https://github.com/jetpacapp/DeepBeliefSDK/issues/57) of their repository. In the meantime I'm working from an alternate base implementation.

**Comment by waylonflinn** [comment]
Wanted to post a quick update here, for interested parties. I've fully converted the jetpac demo to use weblas, producing close to a 2x speedup. The `GEMM` function, in particular, is about 4x faster.

The key to this speedup was a combination of the factors that @mdda outlines above. Namely: transposing the second matrix, full packing of RGBA elements and using a single GPU instruction (`dot`) to process four elements at a time.

A good next step here might be to convert the work @karpathy links to above to make use of this, and see if more impressive speedups ensue. If no one else steps in to take up the torch, I'll likely start on it in the next few days.

**Comment by RangerMauve** [comment]
There's been a release of [a library](http://gpu.rocks/) for converting JS functions to GPU calculations with fallbacks to JS which could make it pretty easy to have complex matrix math done on the GPU while not having to worry about doing it in both GLSL and JS.

**Comment by haehn** [comment]
hi guys, i am really interested in this. what is the current state?

**Comment by jonygli** [comment]
any update?

**Comment by waylonflinn** [comment]
Hey everyone. I'm about to resume development on my convolutional neural net library powered by weblas (an in browser computation library powered by the GPU). Weblas was about five to ten times faster than gpu.js on the relevant functions last time I checked.

Library is here: https://github.com/waylonflinn/webnn

**Comment by soswow** [comment]
I've made some experiments with network I have. Numbers are confusing:
```
Chrome without webgl: 44ms/tick
Firefox without webgl: 100ms/tick
Chrome with webgl: 72ms/tick
Firefox with webgl: 42ms/tick
```
o_O
Where 'tick' is essentially a little bit of game tick logic and two forward passes for each bot.
